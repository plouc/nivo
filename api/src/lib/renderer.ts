import { pick } from 'lodash'
import React, { FunctionComponent } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { ChartProps, chartsMapping, ChartType } from '../mapping'

const staticProps = {
    animate: false,
    isInteractive: false,
    renderWrapper: false,
    theme: {},
}

export const renderChart = <T extends ChartType>(
    {
        type,
        props,
    }: {
        type: T
        props: ChartProps<T>
    },
    override
) => {
    const chart = chartsMapping[type]
    const component = chart.component as FunctionComponent<ChartProps<T>>
    const mergedProps = {
        ...staticProps,
        ...chart.defaults,
        ...props,
        ...pick(override, chart.runtimeProps || []),
    }
    const rendered = renderToStaticMarkup(
        React.createElement<ChartProps<T>>(component, mergedProps)
    )

    return `<?xml version="1.0" ?>${rendered}`
}
