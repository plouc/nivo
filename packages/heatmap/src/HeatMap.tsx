import { ReactNode, Fragment, createElement, useMemo } from 'react'
import { SvgWrapper, Container, useDimensions } from '@nivo/core'
import { Axes, Grid } from '@nivo/axes'
import { AnchoredContinuousColorsLegendSvg } from '@nivo/legends'
import {
    DefaultHeatMapDatum,
    HeatMapDatum,
    HeatMapCommonProps,
    HeatMapSvgProps,
    LayerId,
    CustomLayerProps,
} from './types'
import { useHeatMap } from './hooks'
import { svgDefaultProps } from './defaults'
import { HeatMapCells } from './HeatMapCells'
import { HeatMapCellAnnotations } from './HeatMapCellAnnotations'

type InnerHeatMapProps<Datum extends HeatMapDatum, ExtraProps extends object> = Omit<
    HeatMapSvgProps<Datum, ExtraProps>,
    'animate' | 'motionConfig' | 'renderWrapper' | 'theme'
>

const InnerHeatMap = <Datum extends HeatMapDatum, ExtraProps extends object>({
    data,
    layers = svgDefaultProps.layers,
    valueFormat,
    width,
    height,
    margin: partialMargin,
    forceSquare = svgDefaultProps.forceSquare,
    xInnerPadding = svgDefaultProps.xInnerPadding,
    xOuterPadding = svgDefaultProps.xOuterPadding,
    yInnerPadding = svgDefaultProps.yInnerPadding,
    yOuterPadding = svgDefaultProps.yOuterPadding,
    sizeVariation = svgDefaultProps.sizeVariation,
    cellComponent = svgDefaultProps.cellComponent as NonNullable<
        HeatMapSvgProps<Datum, ExtraProps>['cellComponent']
    >,
    opacity = svgDefaultProps.opacity,
    activeOpacity = svgDefaultProps.activeOpacity,
    inactiveOpacity = svgDefaultProps.inactiveOpacity,
    borderRadius = svgDefaultProps.borderRadius,
    borderWidth = svgDefaultProps.borderWidth,
    borderColor = svgDefaultProps.borderColor as HeatMapCommonProps<Datum>['borderColor'],
    enableGridX = svgDefaultProps.enableGridX,
    enableGridY = svgDefaultProps.enableGridY,
    axisTop = svgDefaultProps.axisTop,
    axisRight = svgDefaultProps.axisRight,
    axisBottom = svgDefaultProps.axisBottom,
    axisLeft = svgDefaultProps.axisLeft,
    enableLabels = svgDefaultProps.enableLabels,
    label = svgDefaultProps.label as HeatMapCommonProps<Datum>['label'],
    labelTextColor = svgDefaultProps.labelTextColor as HeatMapCommonProps<Datum>['labelTextColor'],
    colors = svgDefaultProps.colors as HeatMapCommonProps<Datum>['colors'],
    emptyColor = svgDefaultProps.emptyColor,
    legends = svgDefaultProps.legends,
    annotations = svgDefaultProps.annotations as HeatMapCommonProps<Datum>['annotations'],
    isInteractive = svgDefaultProps.isInteractive,
    onMouseEnter,
    onMouseMove,
    onMouseLeave,
    onClick,
    hoverTarget = svgDefaultProps.hoverTarget,
    tooltip = svgDefaultProps.tooltip as HeatMapCommonProps<Datum>['tooltip'],
    role,
    ariaLabel,
    ariaLabelledBy,
    ariaDescribedBy,
}: InnerHeatMapProps<Datum, ExtraProps>) => {
    const {
        margin: _margin,
        innerWidth: _innerWidth,
        innerHeight: _innerHeight,
        outerWidth,
        outerHeight,
    } = useDimensions(width, height, partialMargin)

    const {
        width: innerWidth,
        height: innerHeight,
        offsetX,
        offsetY,
        xScale,
        yScale,
        cells,
        colorScale,
        activeCell,
        setActiveCell,
    } = useHeatMap<Datum, ExtraProps>({
        data,
        valueFormat,
        width: _innerWidth,
        height: _innerHeight,
        forceSquare,
        xInnerPadding,
        xOuterPadding,
        yInnerPadding,
        yOuterPadding,
        sizeVariation,
        colors,
        emptyColor,
        opacity,
        activeOpacity,
        inactiveOpacity,
        borderColor,
        label,
        labelTextColor,
        hoverTarget,
    })

    const margin = useMemo(
        () => ({
            ..._margin,
            top: _margin.top + offsetY,
            left: _margin.left + offsetX,
        }),
        [_margin, offsetX, offsetY]
    )

    const layerById: Record<LayerId, ReactNode> = {
        grid: null,
        axes: null,
        cells: null,
        legends: null,
        annotations: null,
    }

    if (layers.includes('grid')) {
        layerById.grid = (
            <Grid
                key="grid"
                width={innerWidth} // - offsetX * 2
                height={innerHeight} // - offsetY * 2
                xScale={enableGridX ? xScale : null}
                yScale={enableGridY ? yScale : null}
            />
        )
    }

    if (layers.includes('axes')) {
        layerById.axes = (
            <Axes
                key="axes"
                xScale={xScale}
                yScale={yScale}
                width={innerWidth} // - offsetX * 2
                height={innerHeight} // - offsetY * 2
                top={axisTop}
                right={axisRight}
                bottom={axisBottom}
                left={axisLeft}
            />
        )
    }

    if (layers.includes('cells')) {
        layerById.cells = (
            <Fragment key="cells">
                <HeatMapCells<Datum, ExtraProps>
                    cells={cells}
                    cellComponent={cellComponent}
                    borderRadius={borderRadius}
                    borderWidth={borderWidth}
                    isInteractive={isInteractive}
                    setActiveCell={setActiveCell}
                    onMouseEnter={onMouseEnter}
                    onMouseMove={onMouseMove}
                    onMouseLeave={onMouseLeave}
                    onClick={onClick}
                    tooltip={tooltip}
                    enableLabels={enableLabels}
                />
            </Fragment>
        )
    }

    if (layers.includes('legends') && colorScale !== null) {
        layerById.legends = (
            <Fragment key="legends">
                {legends.map((legend, index) => (
                    <AnchoredContinuousColorsLegendSvg
                        {...legend}
                        key={index}
                        containerWidth={innerWidth}
                        containerHeight={innerHeight}
                        scale={colorScale}
                    />
                ))}
            </Fragment>
        )
    }

    if (layers.includes('annotations') && annotations.length > 0) {
        layerById.annotations = (
            <HeatMapCellAnnotations<Datum>
                key="annotations"
                cells={cells}
                annotations={annotations}
            />
        )
    }

    const customLayerProps: CustomLayerProps<Datum> = {
        cells,
        activeCell,
        setActiveCell,
    }

    return (
        <SvgWrapper
            width={outerWidth}
            height={outerHeight}
            margin={Object.assign({}, margin, {
                top: margin.top, //+ offsetY,
                left: margin.left, // + offsetX,
            })}
            role={role}
            ariaLabel={ariaLabel}
            ariaLabelledBy={ariaLabelledBy}
            ariaDescribedBy={ariaDescribedBy}
        >
            {layers.map((layer, i) => {
                if (typeof layer === 'function') {
                    return <Fragment key={i}>{createElement(layer, customLayerProps)}</Fragment>
                }

                return layerById?.[layer] ?? null
            })}
        </SvgWrapper>
    )
}

export const HeatMap = <
    Datum extends HeatMapDatum = DefaultHeatMapDatum,
    ExtraProps extends object = Record<string, never>
>({
    isInteractive = svgDefaultProps.isInteractive,
    animate = svgDefaultProps.animate,
    motionConfig = svgDefaultProps.motionConfig,
    theme,
    renderWrapper,
    ...otherProps
}: HeatMapSvgProps<Datum, ExtraProps>) => (
    <Container
        {...{
            animate,
            isInteractive,
            motionConfig,
            renderWrapper,
            theme,
        }}
    >
        <InnerHeatMap<Datum, ExtraProps> isInteractive={isInteractive} {...otherProps} />
    </Container>
)
