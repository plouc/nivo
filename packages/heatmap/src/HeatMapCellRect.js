/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react'
import PropTypes from 'prop-types'
import pure from 'recompose/pure'
import { themePropType } from '@nivo/core'

const style = { cursor: 'pointer' }

const HeatMapCellRect = ({
    data,
    value,
    x,
    y,
    width,
    height,
    color,
    opacity,
    borderWidth,
    borderColor,
    enableLabel,
    textColor,
    onHover,
    onLeave,
    onClick,
    theme,
}) => (
    <g
        transform={`translate(${x}, ${y})`}
        onMouseEnter={onHover}
        onMouseMove={onHover}
        onMouseLeave={onLeave}
        onClick={e => {
            onClick(data, e)
        }}
        style={style}
    >
        <rect
            x={width * -0.5}
            y={height * -0.5}
            width={width}
            height={height}
            fill={color}
            fillOpacity={opacity}
            strokeWidth={borderWidth}
            stroke={borderColor}
            strokeOpacity={opacity}
        />
        {enableLabel && (
            <text
                alignmentBaseline="central"
                textAnchor="middle"
                style={{
                    ...theme.labels.text,
                    fill: textColor,
                }}
                fillOpacity={opacity}
            >
                {value}
            </text>
        )}
    </g>
)

HeatMapCellRect.propTypes = {
    data: PropTypes.object.isRequired,
    value: PropTypes.number.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    color: PropTypes.string.isRequired,
    opacity: PropTypes.number.isRequired,
    borderWidth: PropTypes.number.isRequired,
    borderColor: PropTypes.string.isRequired,
    enableLabel: PropTypes.bool.isRequired,
    textColor: PropTypes.string.isRequired,
    onHover: PropTypes.func.isRequired,
    onLeave: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired,
    theme: themePropType.isRequired,
}

export default pure(HeatMapCellRect)
