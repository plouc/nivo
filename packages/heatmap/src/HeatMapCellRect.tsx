import { memo, useMemo } from 'react'
import { animated, to } from '@react-spring/web'
import { useTheme } from '@nivo/core'
import { CellComponentProps, HeatMapDatum } from './types'

const NonMemoizedHeatMapCellRect = <Datum extends HeatMapDatum>({
    cell,
    borderWidth,
    borderRadius,
    animatedProps,
    onMouseEnter,
    onMouseMove,
    onMouseLeave,
    onClick,
    enableLabels,
}: CellComponentProps<Datum>) => {
    const theme = useTheme()

    const handlers = useMemo(
        () => ({
            onMouseEnter: onMouseEnter ? onMouseEnter(cell) : undefined,
            onMouseMove: onMouseMove ? onMouseMove(cell) : undefined,
            onMouseLeave: onMouseLeave ? onMouseLeave(cell) : undefined,
            onClick: onClick ? onClick(cell) : undefined,
        }),
        [cell, onMouseEnter, onMouseMove, onMouseLeave, onClick]
    )

    return (
        <animated.g
            data-testid={`cell.${cell.id}`}
            style={{ cursor: 'pointer' }}
            opacity={animatedProps.opacity}
            {...handlers}
            transform={to(
                [animatedProps.x, animatedProps.y, animatedProps.scale],
                (x, y, scale) => `translate(${x}, ${y}) scale(${scale})`
            )}
        >
            <animated.rect
                transform={to(
                    [animatedProps.width, animatedProps.height],
                    (width, height) => `translate(${width * -0.5}, ${height * -0.5})`
                )}
                key={cell.id}
                fill={animatedProps.color}
                width={animatedProps.width}
                height={animatedProps.height}
                stroke={animatedProps.borderColor}
                strokeWidth={borderWidth}
                rx={borderRadius}
                ry={borderRadius}
            />
            {enableLabels && (
                <animated.text
                    textAnchor="middle"
                    dominantBaseline="central"
                    fill={animatedProps.labelTextColor}
                    style={{
                        ...theme.labels.text,
                        fill: undefined,
                        userSelect: 'none',
                    }}
                >
                    {cell.label}
                </animated.text>
            )}
        </animated.g>
    )
}

export const HeatMapCellRect = memo(NonMemoizedHeatMapCellRect) as typeof NonMemoizedHeatMapCellRect
