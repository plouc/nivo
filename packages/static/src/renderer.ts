import pick from 'lodash/pick'
import { createElement } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { ChartProps, chartsMapping, ChartType } from './mappings'

const staticProps = {
    animate: false,
    isInteractive: false,
    renderWrapper: false,
    theme: {},
}

export const renderChart = <T extends ChartType>(
    {
        type,
        props,
    }: {
        type: T
        props: ChartProps<T>
    },
    override: Partial<ChartProps<T>>
) => {
    const chart = chartsMapping[type]
    const component = chart.component
    const mergedProps = {
        ...staticProps,
        ...chart.defaults,
        ...props,
        ...pick(override, chart.runtimeProps || []),
    }
    const rendered = renderToStaticMarkup(
        // @ts-ignore
        createElement(component, mergedProps)
    )

    return `<?xml version="1.0" ?>${rendered}`
}
