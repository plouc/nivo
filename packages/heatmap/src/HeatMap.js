/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React, { Component } from 'react'
import partial from 'lodash/partial'
import { TransitionMotion } from 'react-motion'
import { colorMotionSpring, getInterpolatedColor, Container, SvgWrapper, Grid } from '@nivo/core'
import { Axes } from '@nivo/axes'
import setDisplayName from 'recompose/setDisplayName'
import { HeatMapPropTypes } from './props'
import computeNodes from './computeNodes'
import enhance from './enhance'
import HeatMapCellRect from './HeatMapCellRect'
import HeatMapCellCircle from './HeatMapCellCircle'
import HeatMapCellTooltip from './HeatMapCellTooltip'

class HeatMap extends Component {
    static propTypes = HeatMapPropTypes

    handleNodeHover = (showTooltip, node, event) => {
        const { setCurrentNode, theme, tooltipFormat, tooltip } = this.props
        setCurrentNode(node)
        showTooltip(
            <HeatMapCellTooltip
                node={node}
                theme={theme}
                format={tooltipFormat}
                tooltip={tooltip}
            />,
            event
        )
    }

    handleNodeLeave = hideTooltip => {
        this.props.setCurrentNode(null)
        hideTooltip()
    }

    render() {
        const {
            xScale,
            yScale,
            offsetX,
            offsetY,

            margin,
            width,
            height,
            outerWidth,
            outerHeight,

            // cells
            cellShape,
            cellBorderWidth,
            getCellBorderColor,

            // axes & grid
            axisTop,
            axisRight,
            axisBottom,
            axisLeft,
            enableGridX,
            enableGridY,

            // labels
            enableLabels,
            getLabelTextColor,

            // theming
            theme,

            // motion
            animate,
            motionStiffness,
            motionDamping,
            boundSpring,

            // interactivity
            isInteractive,
            onClick,
        } = this.props

        let Cell
        if (cellShape === 'rect') {
            Cell = HeatMapCellRect
        } else if (cellShape === 'circle') {
            Cell = HeatMapCellCircle
        } else {
            Cell = cellShape
        }

        const nodes = computeNodes(this.props)

        const motionProps = {
            animate,
            motionDamping,
            motionStiffness,
        }

        return (
            <Container isInteractive={isInteractive} theme={theme}>
                {({ showTooltip, hideTooltip }) => {
                    const onHover = partial(this.handleNodeHover, showTooltip)
                    const onLeave = partial(this.handleNodeLeave, hideTooltip)

                    return (
                        <SvgWrapper
                            width={outerWidth}
                            height={outerHeight}
                            margin={Object.assign({}, margin, {
                                top: margin.top + offsetY,
                                left: margin.left + offsetX,
                            })}
                            theme={theme}
                        >
                            <Grid
                                theme={theme}
                                width={width - offsetX * 2}
                                height={height - offsetY * 2}
                                xScale={enableGridX ? xScale : null}
                                yScale={enableGridY ? yScale : null}
                                {...motionProps}
                            />
                            <Axes
                                xScale={xScale}
                                yScale={yScale}
                                width={width}
                                height={height}
                                theme={theme}
                                top={axisTop}
                                right={axisRight}
                                bottom={axisBottom}
                                left={axisLeft}
                                {...motionProps}
                            />
                            {!animate &&
                                nodes.map(node =>
                                    React.createElement(Cell, {
                                        key: node.key,
                                        data: node,
                                        value: node.value,
                                        x: node.x,
                                        y: node.y,
                                        width: node.width,
                                        height: node.height,
                                        color: node.color,
                                        opacity: node.opacity,
                                        borderWidth: cellBorderWidth,
                                        borderColor: getCellBorderColor(node),
                                        enableLabel: enableLabels,
                                        textColor: getLabelTextColor(node),
                                        onHover: partial(onHover, node),
                                        onLeave,
                                        onClick,
                                        theme,
                                    })
                                )}

                            {animate === true && (
                                <TransitionMotion
                                    styles={nodes.map(node => {
                                        return {
                                            key: node.key,
                                            data: node,
                                            style: {
                                                x: boundSpring(node.x),
                                                y: boundSpring(node.y),
                                                width: boundSpring(node.width),
                                                height: boundSpring(node.height),
                                                opacity: boundSpring(node.opacity),
                                                ...colorMotionSpring(node.color, {
                                                    damping: motionDamping,
                                                    stiffness: motionStiffness,
                                                }),
                                            },
                                        }
                                    })}
                                >
                                    {interpolatedStyles => {
                                        return (
                                            <g>
                                                {interpolatedStyles.map(
                                                    ({ key, style, data: node }) => {
                                                        const color = getInterpolatedColor(style)

                                                        return React.createElement(Cell, {
                                                            key,
                                                            data: node,
                                                            value: node.value,
                                                            x: style.x,
                                                            y: style.y,
                                                            width: Math.max(style.width, 0),
                                                            height: Math.max(style.height, 0),
                                                            color,
                                                            opacity: style.opacity,
                                                            borderWidth: cellBorderWidth,
                                                            borderColor: getCellBorderColor({
                                                                ...node,
                                                                color,
                                                            }),
                                                            enableLabel: enableLabels,
                                                            textColor: getLabelTextColor({
                                                                ...node,
                                                                color,
                                                            }),
                                                            onHover: partial(onHover, node),
                                                            onLeave,
                                                            onClick,
                                                            theme,
                                                        })
                                                    }
                                                )}
                                            </g>
                                        )
                                    }}
                                </TransitionMotion>
                            )}
                        </SvgWrapper>
                    )
                }}
            </Container>
        )
    }
}

export default setDisplayName('HeatMap')(enhance(HeatMap))
