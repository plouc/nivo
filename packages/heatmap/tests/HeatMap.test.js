import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import HeatMapCellRect from '../src/HeatMapCellRect'
import HeatMap from '../src/HeatMap'

const sampleProps = {
    indexBy: 'age',
    keys: ['rent', 'loan', 'income'],
    data: [
        {
            age: 20,
            rent: 30,
            loan: 24,
            income: 60,
        },
        {
            age: 30,
            rent: 40,
            loan: 50,
            income: 90,
        },
        {
            age: 40,
            rent: 20,
            loan: 13,
            income: 40,
        },
    ],
    width: 500,
    height: 300,
    animate: false,
}

it('should render a basic heat map chart', () => {
    const component = renderer.create(<HeatMap {...sampleProps} />)

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

it('should allow to disable labels', () => {
    const wrapper = mount(<HeatMap {...sampleProps} enableLabels={false} />)

    const cells = wrapper.find(HeatMapCellRect)
    expect(cells).toHaveLength(9)
    cells.forEach(cell => {
        expect(cell.find('text').exists()).toBe(false)
    })
})

it('should allow to format labels', () => {
    const wrapper = mount(
        <HeatMap {...sampleProps} label={(datum, key) => datum[key].toFixed(2)} />
    )

    const cells = wrapper.find(HeatMapCellRect)
    expect(cells).toHaveLength(9)
    cells.forEach(cell => {
        expect(cell.find('text').exists()).toBe(true)
        expect(cell.find('text').text().length).toEqual(5)
    })
})
