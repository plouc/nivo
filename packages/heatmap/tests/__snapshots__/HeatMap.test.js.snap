// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a basic heat map chart 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <svg
    height={300}
    role="img"
    width={500}
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      fill="transparent"
      height={300}
      width={500}
    />
    <g
      transform="translate(0,0)"
    >
      <g
        transform="translate(0,0)"
      >
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(83.33333333333333,0)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={-5}
          />
          <text
            dominantBaseline="alphabetic"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="middle"
            transform="translate(0,-10) rotate(0)"
          >
            rent
          </text>
        </g>
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(250,0)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={-5}
          />
          <text
            dominantBaseline="alphabetic"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="middle"
            transform="translate(0,-10) rotate(0)"
          >
            loan
          </text>
        </g>
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(416.66666666666663,0)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={-5}
          />
          <text
            dominantBaseline="alphabetic"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="middle"
            transform="translate(0,-10) rotate(0)"
          >
            income
          </text>
        </g>
        <line
          style={
            Object {
              "stroke": "transparent",
              "strokeWidth": 1,
            }
          }
          x1={0}
          x2={500}
          y1={0}
          y2={0}
        />
      </g>
      <g
        transform="translate(0,0)"
      >
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(0,50)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-5}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="end"
            transform="translate(-10,0) rotate(0)"
          >
            20
          </text>
        </g>
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(0,150)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-5}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="end"
            transform="translate(-10,0) rotate(0)"
          >
            30
          </text>
        </g>
        <g
          style={
            Object {
              "opacity": 1,
            }
          }
          transform="translate(0,250)"
        >
          <line
            style={
              Object {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-5}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              Object {
                "fill": "#333333",
                "fontFamily": "sans-serif",
                "fontSize": 11,
              }
            }
            textAnchor="end"
            transform="translate(-10,0) rotate(0)"
          >
            40
          </text>
        </g>
        <line
          style={
            Object {
              "stroke": "transparent",
              "strokeWidth": 1,
            }
          }
          x1={0}
          x2={0}
          y1={0}
          y2={300}
        />
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(83.33333333333333, 50)"
      >
        <rect
          fill="rgba(244, 117, 96, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(244, 117, 96, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(148, 71, 58, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          30
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(250, 50)"
      >
        <rect
          fill="rgba(215, 100, 69, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(215, 100, 69, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(130, 61, 42, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          24
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(416.66666666666663, 50)"
      >
        <rect
          fill="rgba(151, 227, 213, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(151, 227, 213, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(92, 138, 129, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          60
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(83.33333333333333, 150)"
      >
        <rect
          fill="rgba(232, 193, 160, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(232, 193, 160, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(141, 117, 97, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          40
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(250, 150)"
      >
        <rect
          fill="rgba(232, 193, 160, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(232, 193, 160, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(141, 117, 97, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          50
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(416.66666666666663, 150)"
      >
        <rect
          fill="rgba(0, 176, 167, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(0, 176, 167, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(0, 107, 101, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          90
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(83.33333333333333, 250)"
      >
        <rect
          fill="rgba(215, 100, 69, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(215, 100, 69, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(130, 61, 42, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          20
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(250, 250)"
      >
        <rect
          fill="rgba(215, 100, 69, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(215, 100, 69, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(130, 61, 42, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          13
        </text>
      </g>
      <g
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        transform="translate(416.66666666666663, 250)"
      >
        <rect
          fill="rgba(232, 193, 160, 1)"
          fillOpacity={0.85}
          height={100}
          stroke="rgba(232, 193, 160, 1)"
          strokeOpacity={0.85}
          strokeWidth={0}
          width={166.66666666666666}
          x={-83.33333333333333}
          y={-50}
        />
        <text
          dominantBaseline="central"
          fillOpacity={0.85}
          style={
            Object {
              "fill": "rgba(141, 117, 97, 1)",
              "fontFamily": "sans-serif",
              "fontSize": 11,
            }
          }
          textAnchor="middle"
        >
          40
        </text>
      </g>
    </g>
  </svg>
</div>
`;
